### PR TITLE
Measure uplift from Acceptable Ads

### PR DESCRIPTION
**Pull request by davidfischer**
I don't think this is quite perfect but I wanted to start a discussion. There's two main issues as far as I see.

There's a possible race condition between checking whether ABP allows the ad (loading 2 gifs) and when the ad loads. If the gifs load very slowly, the ad will already be loaded and the viewed -- which happens 100ms after the ad loads -- then it the uplift check won't happen before the view is sent. I hate the alternative though which is to wait for the gifs to load before showing the ad. That will slow down every ad load.

The other issue is that in my testing the test ABP pixel (https://media.ethicalads.io/media/px.gif) is blocked by a more generic filter rule on the easylist (`||ethicalads.io^$third-party`). Since the pixel is a third-party request in this context, it gets blocked so the detection wasn't working. We could put this on a different domain or use a non-CDNed version. I don't love the options here.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -228,6 +281,16 @@ export function load_placements(force_load = false) {
   return Promise.all(
     elements.map((element) => {
       const placement = Placement.from_element(element);
+
+      if (placement && !force_load) {
+        placement.detectABP(ABP_DETECTION_PX, function (usesABP) {
+          placement.uplifted = usesABP;
+          if (usesABP) {
+            console.log("Uses ABP");
```
Comment (ericholscher) [inline index.js:289]: This should probably be more descriptive:

```suggestion
            console.debug("Acceptable Ads enabled. Thanks for allowing our non-tracking ads :)");
```

File: index.js [hunk 2]
```diff
@@ -199,6 +206,52 @@ export class Placement {
...
+
+  /* Detect whether this ad is "uplifted" meaning allowed by ABP's Acceptable Ads list
+   *
+   * Calls the provided callback passing a boolean whether this ad is uplifted.
+   * We need this data to provide back to the AcceptableAds folks.
+   *
+   * @static
+   * @param {string} px - A URL of a pixel to test
+   * @param {function) callback - A callback to call when finished
+   */
```
Comment (ericholscher) [inline index.js:218]: We should probably note this code came via the Eyeo team.

**Review by ericholscher — APPROVED** [review]
Simple enough 👍 

Annoying we can't meaningfully test it, but the code is pretty simple.

**Comment by ericholscher** [comment]
> There's a possible race condition between checking whether ABP allows the ad (loading 2 gifs) and when the ad loads. If the gifs load very slowly, the ad will already be loaded and the viewed -- which happens 100ms after the ad loads -- then it the uplift check won't happen before the view is sent. I hate the alternative though which is to wait for the gifs to load before showing the ad. That will slow down every ad load.

@davidfischer we don't want to wait to show the ad right, just wait to send the "view" pixel? I think we'll have to see how it works in practice, but delaying the pixel view doesn't seem terrible.

**Comment by davidfischer** [comment]
> @davidfischer we don't want to wait to show the ad right, just wait to send the "view" pixel? I think we'll have to see how it works in practice, but delaying the pixel view doesn't seem terrible.

I can see if I can work that. The waiting to load 2 gifs may not prove to be a big deal in practice. In the case ABP + AcceptableAds is running, at least 1 of the gifs will be rejected immediately so it's a timing issue between loading a gif via a CDN and loading an ad from our server. The former should be faster almost always.

**Pull request merged**